shared/devices/container.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+import Render from '.'
+import flags from '../util/feature-flags'
+import {List} from 'immutable'
+import {addNewPhone, addNewComputer} from '../actions/login'
+import {compose, lifecycle, mapProps, withState} from 'recompose'
+import {connect} from 'react-redux'
+import {createSelector} from 'reselect'
+import {load, paperKeyMake} from '../actions/devices'
+
+import type {TypedState} from '../constants/reducer'
+
+const getAllDevicesSelector = (state: TypedState) => state.devices.get('deviceIDs')
+const getDeviceEntitiesSelector = (state: TypedState) => state.entities.get('devices')
+
+const getDevicesAndRevokedDevicesSelector = createSelector(
+  getAllDevicesSelector, getDeviceEntitiesSelector,
+  (allDevices, deviceEntities) => {
+    const split = allDevices.groupBy(id => deviceEntities.get(id).revokedAt ? 'revokedDeviceIDs' : 'deviceIDs')
+    const deviceIDs = split.get('deviceIDs', List())
+    const revokedDeviceIDs = split.get('revokedDeviceIDs', List())
+    return {
+      deviceIDs,
+      revokedDeviceIDs,
+    }
+  }
+)
+
+const mapStateToProps = (state: any, {routeState}) => {
+  const {showingRevoked} = routeState
+  const {deviceIDs, revokedDeviceIDs} = getDevicesAndRevokedDevicesSelector(state)
+
+  return {
+    deviceIDs,
+    revokedDeviceIDs,
+    showingRevoked,
+  }
+}
+
+const mapDispatchToProps = (dispatch: any, {routeState, setRouteState}) => ({
+  addNewComputer: () => dispatch(addNewComputer()),
+  addNewPaperKey: () => dispatch(paperKeyMake()),
+  addNewPhone: () => dispatch(addNewPhone()),
+  loadDevices: () => dispatch(load()),
+  onToggleShowRevoked: () => { setRouteState({showingRevoked: !routeState.showingRevoked}) },
+})
+
+const menuItems = props => ([
+  ...flags.mobileAppsExist && [{onClick: props.addNewPhone, title: 'New Phone'}] || [],
+  {onClick: props.addNewComputer, title: 'New computer'},
+  {onClick: props.addNewPaperKey, title: 'New paper key'},
+])
+
+const Devices = compose(
+  lifecycle({
+    componentWillMount: function () {
+      this.props.loadDevices()
+    },
+  }),
+  // Don't pass immutable things to dumb components
+  mapProps(props => ({
```

Comment:
I didn't touch the functionality of the render functions that much. Mostly hoisted up some props and reduced some code duplication. In general I think the dumb components should be a lot dumber than they are now. They should always take simple json props and get flat props that they consume simply. Dumb components knowing deeply about what a message is and types and things like that makes them unportable